### PR TITLE
chore: send email to email receiver when tasks are finished successfully

### PR DESCRIPTION
**Pull request by Dinasoa**

### CODE REVIEW

File: src/main/java/school/hei/geotiler/service/ZoneTilingTaskService.java [hunk 1]
```diff
@@ -5,19 +5,30 @@
...
 import school.hei.geotiler.repository.model.ZoneTilingTask;
+import school.hei.geotiler.utils.HTMLUtils;
 
 @Service
 @AllArgsConstructor
 public class ZoneTilingTaskService {
   private final ZoneTilingTaskRepository repository;
   private final ZoneTilingJobService zoneTilingJobService;
+  private final Mailer mailer;
+  public static final String EMAIL_OBJECT = "Resultat du traitement de zone";
```
Comment (hei-teacher) [inline src/main/java/school/hei/geotiler/service/ZoneTilingTaskService.java:30]: Why "Traitement de zone" while the name of the class is ZoneTiling, not ZoneTreatment? Be precise!

Also, Resultat --> Résultat

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/geotiler/utils/HTMLUtils.java [hunk 2]
```diff
@@ -0,0 +1,25 @@
+package school.hei.geotiler.utils;
+
+import org.thymeleaf.TemplateEngine;
+import org.thymeleaf.context.Context;
+import org.thymeleaf.templatemode.TemplateMode;
+import org.thymeleaf.templateresolver.ClassLoaderTemplateResolver;
+
+public class HTMLUtils {
```
Comment (hei-teacher) [inline src/main/java/school/hei/geotiler/utils/HTMLUtils.java:8]: Never ever use Utils againa. @Mahefaa > Explain her

File: src/main/java/school/hei/geotiler/utils/HTMLUtils.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  public static String parseToString(String template, Context context) {
+    TemplateEngine engine = configureEngine();
+    return engine.process(template, context);
+  }
+
+  private static TemplateEngine configureEngine() {
+    ClassLoaderTemplateResolver templateResolver = new ClassLoaderTemplateResolver();
```
Comment (hei-teacher) [inline src/main/java/school/hei/geotiler/utils/HTMLUtils.java:15]: Use var when the type is already explicited by new

**Review by hei-teacher — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/geotiler/utils/HTMLUtils.java (hunk 2 above)
Comment (Dinasoa) [reply src/main/java/school/hei/geotiler/utils/HTMLUtils.java:8]: Waiting for his explaination

File: src/main/resources/templates/zone_tiling.html [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+        body {
+            font-family: Helvetica, serif;
+        }
+    </style>
+</head>
+<body>
+<section>
+    <p>Bonjour,</p>
+    <p th:text="${zone}">
+        Le pavage de la zone: [[${zone}]] a été effectuée avec succes.
```
Comment (Mahefaa) [inline src/main/resources/templates/zone_tiling.html:13]: succès

File: src/main/java/school/hei/geotiler/service/ZoneTilingTaskService.java [hunk 5]
```diff
@@ -44,7 +56,22 @@ public ZoneTilingTask finishWithSuccess(ZoneTilingTask zoneTilingTask) {
...
+            new Email(
+                new InternetAddress(associatedJob.getEmailReceiver()),
+                List.of(),
+                List.of(),
+                EMAIL_OBJECT,
+                emailBody,
+                List.of()));
+      } catch (AddressException e) {
+        throw new ApiException(ApiException.ExceptionType.SERVER_EXCEPTION, e);
+      }
```
Comment (Mahefaa) [inline src/main/java/school/hei/geotiler/service/ZoneTilingTaskService.java:74]: these actions do not belong here

**Review by Mahefaa — CHANGES_REQUESTED** [review]
don't forget to format your code

File: src/main/java/school/hei/geotiler/service/EmailService.java [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+  private final Mailer mailer;
+  public static final String EMAIL_OBJECT = "Résultat du pavage de la zone";
+  public static final String ZONE_TILING_TEMPLATE_NAME = "zone_tiling";
+  private HTMLTemplateParser htmlTemplateParser;
+
+  @SneakyThrows
+  public void sendEmail(ZoneTilingJob zoneTilingJob) {
+    Context context = new Context();
+    context.setVariable("zone", zoneTilingJob.getZoneName());
+    String emailBody = htmlTemplateParser.apply(ZONE_TILING_TEMPLATE_NAME,context);
```
Comment (Mahefaa) [inline src/main/java/school/hei/geotiler/service/EmailService.java]: missing space between the two function args ( separate first arg + comma from second arg)

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/geotiler/service/api/TilesDownloaderApi.java [hunk 7]
```diff
@@ -33,16 +34,18 @@ public byte[] downloadTiles(Parcel parcel) {
...
-        .queryParam("zoom_size", parcel.getFeature().getZoom());
+    UriComponentsBuilder builder =
+        UriComponentsBuilder.fromHttpUrl(geoTilesDownloaderApiURl)
+            .queryParam("zoom_size", parcel.getFeature().getZoom());
 
-    ResponseEntity<byte[]> responseEntity = restTemplate.postForEntity(builder.toUriString(), request, byte[].class);
+    ResponseEntity<byte[]> responseEntity =
+        restTemplate.postForEntity(builder.toUriString(), request, byte[].class);
 
     return responseEntity.getBody();
```
Comment (Mahefaa) [inline src/main/java/school/hei/geotiler/service/api/TilesDownloaderApi.java:44]: check the statusCode of the response and throw ApiException(ServerException) if not 200

**Review by Mahefaa — CHANGES_REQUESTED** [review]

File: src/main/java/school/hei/geotiler/service/api/TilesDownloaderApi.java [hunk 8]
```diff
@@ -33,16 +37,22 @@ public byte[] downloadTiles(Parcel parcel) {
...
+            .queryParam("zoom_size", parcel.getFeature().getZoom());
+
+    ResponseEntity<byte[]> responseEntity =
+        restTemplate.postForEntity(builder.toUriString(), request, byte[].class);
+
+    if (responseEntity.getStatusCode().value() == 200) {
+      return responseEntity.getBody();
+    } else {
+      throw new ApiException(SERVER_EXCEPTION, "Server error");
+    }
```
Comment (Mahefaa) [inline src/main/java/school/hei/geotiler/service/api/TilesDownloaderApi.java:51]: complex if else

**Review by Mahefaa — CHANGES_REQUESTED** [review]

**Review by hei-teacher — APPROVED** [review]

**Review by Mahefaa — APPROVED** [review]
LGTM

**Pull request merged**